nvflare/app_common/hub/hub_controller.py

```diff
@@ -0,0 +1,490 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import time
+from typing import Union
+
+from nvflare.apis.client import Client
+from nvflare.apis.controller_spec import (
+    ClientTask,
+    ControllerSpec,
+    OperatorConfigKey,
+    OperatorMethod,
+    Task,
+    TaskOperatorKey,
+)
+from nvflare.apis.event_type import EventType
+from nvflare.apis.fl_component import FLComponent
+from nvflare.apis.fl_constant import FLContextKey, ReturnCode
+from nvflare.apis.fl_context import FLContext
+from nvflare.apis.impl.controller import Controller
+from nvflare.apis.operator_spec import OperatorSpec
+from nvflare.apis.shareable import ReservedHeaderKey, Shareable, make_reply
+from nvflare.apis.signal import Signal
+from nvflare.app_common.abstract.aggregator import Aggregator
+from nvflare.app_common.abstract.learnable_persistor import LearnablePersistor
+from nvflare.app_common.abstract.shareable_generator import ShareableGenerator
+from nvflare.app_common.app_constant import AppConstants
+from nvflare.app_common.app_event_type import AppEventType
+from nvflare.fuel.utils.pipe.pipe import Pipe
+from nvflare.fuel.utils.pipe.pipe_monitor import PipeMonitor, Topic
+from nvflare.fuel.utils.validation_utils import check_object_type, check_positive_number, check_str
+
+
+class BcastOperator(OperatorSpec, FLComponent):
+
+    _PROP_AGGR = "aggr"
+
+    def __init__(self):
+        OperatorSpec.__init__(self)
+        FLComponent.__init__(self)
+        self.current_aggregator = None
+
+    def operate(
+        self,
+        op_description: dict,
+        controller: ControllerSpec,
+        task_name: str,
+        task_data: Shareable,
+        abort_signal: Signal,
+        fl_ctx: FLContext,
+    ) -> Union[Shareable, None]:
+        aggr_id = op_description.get(TaskOperatorKey.AGGREGATOR, "")
+        if not aggr_id:
+            self.log_error(fl_ctx, "missing aggregator component id")
+            return None
+
+        engine = fl_ctx.get_engine()
+        aggr = engine.get_component(aggr_id)
+        if not aggr:
+            self.log_error(fl_ctx, f"no aggregator defined for component id {aggr_id}")
+            return None
+
+        if not isinstance(aggr, Aggregator):
+            self.log_error(fl_ctx, f"component {aggr_id} must be Aggregator but got {type(aggr)}")
+            return None
+
+        # reset the internal state of the aggregator for next round of aggregation
+        self.current_aggregator = aggr
+        aggr.reset(fl_ctx)
+
+        total_num_clients = len(engine.get_clients())
+        timeout = op_description.get(TaskOperatorKey.TIMEOUT, 0)
+        wait_time_after_min_resps = op_description.get(TaskOperatorKey.WAIT_TIME_AFTER_MIN_RESPS, 5)
+        min_clients = op_description.get(TaskOperatorKey.MIN_TARGETS, 0)
+        if min_clients > total_num_clients:
+            min_clients = total_num_clients
+            wait_time_after_min_resps = 0
```

Comment:
Why do we do this change?
If min clients is greater than total_num_clients we should NOT run this job right?